### PR TITLE
ADAPT - converted to es6 and did some tidy up around show correct answer

### PR DESCRIPTION
**Pull request by eleanor-heath**
Contributing towards OS code: adaptlearning/adapt_framework#2808
"Convert all core components to ES6 classes #2808"

Converted to es6 and did some tidy up around show correct answer.

### CODE REVIEW

**Review by cahirodoherty-learningpool — APPROVED** [review]
👀

File: js/sliderView.js [hunk 1]
```diff
@@ -180,107 +182,107 @@ define([
...
+      this.$('.slider-bar').animate({ width: '0px' });
       this.setSliderValue(this.model.get('_items')[0].value);
-    },
+    }
 
-    onCannotSubmit: function() {},
+    onCannotSubmit() { }
 
-    setSliderValue: function (value) {
+    setSliderValue(value) {
```
Comment (oliverfoster) [inline js/sliderView.js:205]: This function is an example of an easy "return early".
The main body of this function is:
```js
this.$slider.val(value).change();
```

The main body should only be executed if the given condition is met:
```js
if (this.$slider) { }
```
If we therefore invert the condition and return early, we can reduce the nesting and reduce number of lines whilst increasing the clarity:
```js
if (!this.$slider) return;
this.$slider.val(value).change();
```

File: js/sliderView.js [hunk 2]
```diff
@@ -289,74 +291,77 @@ define([
...
-    },
+    }
 
-    showCorrectAnswer: function() {
-      var answers = [];
+    showCorrectAnswer() {
+      const answers = [];
+      this.showScaleMarker(false);
 
+      //are we dealing with a single correct answer or a range?
```
Comment (oliverfoster) [inline js/sliderView.js:300]: Double existence comparison can be simplified:
```js
      const correctAnswer = this.model.get('_correctAnswer');
      if (correctAnswer !== undefined) {
          answers.push(correctAnswer);
      }
```

**Review by ethan-lp — APPROVED** [review]
👀

File: js/sliderView.js [hunk 3]
```diff
@@ -289,74 +291,77 @@ define([
...
-      //are we dealing with a single correct answer or a range?
-      if (correctAnswer) {
-        answers.push(correctAnswer);
-      } else if (bottom !== undefined && top !== undefined) {
-        var answer = this.model.get('_correctRange')._bottom;
-        var topOfRange = this.model.get('_correctRange')._top;
-        while (answer <= topOfRange) {
-          answers.push(answer);
-          answer += step;
+      if (answers.length === 0) {
```
Comment (oliverfoster) [inline js/sliderView.js:309]: Use variables to increase clarity and reduce nesting. Apply "return early" principle.
```js
      const correctRange = this.model.has('_correctRange');
      const hasCorrectRange = (correctRange !== undefined);

      if (!hasCorrectAnswer && !hasCorrectRange) {
        console.log('adapt-contrib-slider::WARNING: no correct answer or correct range set in JSON');
      }
      
      if (!hasCorrectAnswer && hasCorrectRange) {
        const bottom = correctRange ._bottom;
        const top =correctRange ._top;
        const step = (this.model.has('_scaleStep') ? this.model.get('_scaleStep') : 1);

        if (bottom !== undefined && top !== undefined) {
          const answer = this.model.get('_correctRange')._bottom;
          const topOfRange = this.model.get('_correctRange')._top;
          while (answer <= topOfRange) {
            answers.push(answer);
            answer += step;
          }
        }
      }
```

File: js/sliderView.js [hunk 4]
```diff
@@ -365,26 +370,26 @@ define([
...
 
     // this makes the marker visible or hidden
-    showScaleMarker: function(show) {
-      var $scaleMarker = this.$('.js-slider-number-selection');
+    showScaleMarker(show) {
+      const $scaleMarker = this.$('.js-slider-number-selection');
       if (this.model.get('_showScaleIndicator')) {
         this.showNumber(show);
-        if(show) {
+        if (show) {
```
Comment (oliverfoster) [inline js/sliderView.js:382]:
```js
$scaleMarker.toggleClass('display-block', show);
```

File: js/sliderView.js [hunk 5]
```diff
@@ -180,107 +182,107 @@ define([
...
+        const $number = $numbers.eq(index);
+        const newLeft = Math.round($number.data('normalisedPosition') * scaleWidth);
+        $number.css({ left: newLeft });
       });
-    },
+    }
 
     //Labels are enabled in slider.hbs. Here we manage their containing div.
-    showLabels: function () {
+    showLabels() {
```
Comment (oliverfoster) [inline js/sliderView.js:270]: Return early.

File: js/sliderView.js [hunk 6]
```diff
@@ -289,74 +291,77 @@ define([
...
       this.animateToPosition(this.mapIndexToPixels(userAnswerIndex));
       this.setSliderValue(this.model.get('_userAnswer'));
-    },
+    }
 
     // according to given item index this should make the item as selected
-    selectItem: function(itemIndex, noFocus) {
-      _.each(this.model.get('_items'), function(item, index) {
+    selectItem(itemIndex) {
+      this.model.get('_items').forEach((item, index) => {
```
Comment (tomgreenfield) [inline js/sliderView.js:364]: Could refactor to remove the loop:
```js
const item = this.model.get('_items')[itemIndex];
if (!item) return;
item.selected = true;
this.model.set('_selectedItem', item);
this.$('.js-slider-item-input').attr({
  'value': item.value,
  'aria-valuenow': item.value
});
```

File: js/sliderView.js (hunk 3 above)
Comment (eleanor-heath) [reply js/sliderView.js:309]: all great suggestions thanks, addressing them currently.

would you oppose if I was to make `if (!hasCorrectAnswer && hasCorrectRange) {`
`else if (hasCorrectRange) {` ?

**Review by tomgreenfield — COMMENTED** [review]
:eye:

File: js/sliderView.js (hunk 3 above)
Comment (oliverfoster) [reply js/sliderView.js:309]: Nope I wouldn't oppose, that seems fine.

Comment (tomgreenfield) [reply js/sliderView.js:309]: Thanks @oliverfoster for the suggestion – the more I read the code, the more I feel there's enough going on to warrant a separate function. Something like:
```js
getCorrectAnswers() {
  const answerSingle = this.model.get('_correctAnswer');
  if (answerSingle) {
    return [ answerSingle ];
  }
  const answerMultiple = this.model.get('_correctRange');
  const answers = [];
  if (!answerMultiple) {
    console.log('adapt-contrib-slider::WARNING: no correct answer or correct range set in JSON');
    return answers;
  }
  const bottom = answerMultiple._bottom;
  const top = answerMultiple._top;
  if (bottom === undefined || top === undefined) {
    return answers;
  }
  let answer = bottom;
  const step = this.model.get('_scaleStep') || 1;
  while (answer <= top) {
    answers.push(answer);
    answer += step;
  }
  return answers;
}
```

Comment (oliverfoster) [reply js/sliderView.js:309]: Yea, that block in the middle does seem somewhat out of place.

Comment (eleanor-heath) [reply js/sliderView.js:309]: I like this idea - splits the code up a bit meaning you have a show and a get function 👍 

I was also thinking that we should only attempt to animate to position/show model answer and set slider value when there are answers. This means we could return after logging out the warning when there are no answers.

For example, 

`const answers = this.getCorrectAnswers();
  if (!answers.length) {
    console.log('adapt-contrib-slider::WARNING: no correct answer or correct range set in JSON');
    return
  } else {
    const middleAnswer = answers[Math.floor(answers.length / 2)];
    this.animateToPosition(this.mapIndexToPixels(this.getIndexFromValue(middleAnswer)));
    this.showModelAnswers(answers);
    this.setSliderValue(middleAnswer);
  }
` 

Let me know your thoughts!

File: js/sliderView.js (hunk 6 above)
Comment (eleanor-heath) [reply js/sliderView.js]: I have resolved all of your comments. Just a note on this, the first item is defaulted to be selected so have called on deselectAllItems before I select the chosen item to be sure there is not anything currently selected. If we dont do this the value of isCorrect could be wrong, so just to be sure I want to deselect them all and not just the first item.

**Comment by eleanor-heath** [comment]
@tomgreenfield @oliverfoster I have addressed all of your comments, can you take another look please when you have the chance.

File: js/sliderModel.js [hunk 7]
```diff
@@ -70,84 +69,78 @@ define([
...
+    }
 
-      _.each(this.get('_items'), function(item, index) {
+    deselectAllItems() {
+      this.get('_items').forEach(item => item.selected = false);
+    }
+    
+    isCorrect() {
+      let numberOfCorrectAnswers = 0;
+      this.get('_items').forEach(item => {
```
Comment (oliverfoster) [inline js/sliderModel.js:105]:
```suggestion
      this.set('_isAtLeastOneCorrectSelection', false);
      this.get('_items').forEach(item => {
```

File: js/sliderModel.js [hunk 8]
```diff
@@ -70,84 +69,78 @@ define([
...
         if (item.selected && item.correct) {
           this.set('_isAtLeastOneCorrectSelection', true);
           numberOfCorrectAnswers++;
         }
-      }, this);
+      });
 
       this.set('_numberOfCorrectAnswers', numberOfCorrectAnswers);
 
       return this.get('_isAtLeastOneCorrectSelection') ? true : false;
```
Comment (oliverfoster) [inline js/sliderModel.js:114]:
```suggestion
      return this.get('_isAtLeastOneCorrectSelection');
```

File: js/sliderView.js [hunk 9]
```diff
@@ -180,107 +177,105 @@ define([
...
+    showScale() {
+      const $markers = this.$('.js-slider-scale-notch-container').empty();
 
       if (this.model.get('_showScale') === false) {
-        $markers.eq(0).css({display: 'none'});
-        this.$('.js-slider-number').css(
-          this.model.get('_showScaleIndicator') ? {visibility: 'hidden'} : {display: 'none'}
+        $markers.eq(0).addClass('u-display-none');
+        this.$('.js-slider-number').addClass(
+          this.model.get('_showScaleIndicator') ? 'visibility-hidden' : 'u-display-none'
```
Comment (oliverfoster) [inline js/sliderView.js:233]:
```suggestion
          this.model.get('_showScaleIndicator') ? 'u-visibility-hidden' : 'u-display-none'
```

File: js/sliderView.js [hunk 10]
```diff
@@ -180,107 +177,105 @@ define([
...
-    },
-
-    remapSliderBar: function() {
-      var $scaler = this.$('.js-slider-scale');
-      var currentIndex = this.getIndexFromValue(this.model.get('_selectedItem').value);
-      var left = this.mapIndexToPixels(currentIndex, $scaler);
-      this.$('.slider-handle').css({left: left + 'px'});
-      this.$('.js-slider-number-selection').css({left: left + 'px'});
+    showLabels() {
+      if (this.model.get('labelStart') && this.model.get('labelEnd')) return;
```
Comment (oliverfoster) [inline js/sliderView.js:265]:
```suggestion
      if (this.model.get('labelStart') || this.model.get('labelEnd')) return;
```
You need to invert the operators as well

File: js/sliderView.js [hunk 11]
```diff
@@ -289,102 +284,101 @@ define([
...
+    }
+
+    showModelAnswers(correctAnswerArray) {
+      const $parentDiv = this.$('.js-slider-model-range');
+      correctAnswerArray.forEach((correctAnswer, index) => {
         $parentDiv.append($('<div class="slider__number-model-answer">'));
 
-        var $element = $(this.$('.js-slider-model-range .slider__number-model-answer')[index]);
-        var startingLeft = this.mapIndexToPixels(this.getIndexFromValue(this.model.get('_selectedItem').value));
+        const $element = $(this.$('.js-slider-model-range .slider__number-model-answer')[index]);
```
Comment (oliverfoster) [inline js/sliderView.js:334]:
```suggestion
        const $element = this.$('.js-slider-model-range .slider__number-model-answer').eq(index);
```

**Review by oliverfoster — APPROVED** [review]

File: js/sliderModel.js (hunk 7 above)
Comment (oliverfoster) [reply js/sliderModel.js:105]: Part 1 to remove the weird `return this.get('_isAtLeastOneCorrectSelection') ? true : false;` ternary

File: js/sliderModel.js (hunk 8 above)
Comment (oliverfoster) [reply js/sliderModel.js:114]: Part 2

File: js/sliderView.js (hunk 9 above)
Comment (oliverfoster) [reply js/sliderView.js:233]: [u-visibility-hidden](https://github.com/adaptlearning/adapt_framework/blob/master/src/core/less/_defaults/utility.less#L32)

File: js/sliderView.js (hunk 11 above)
Comment (oliverfoster) [reply js/sliderView.js:334]: [jQuery.eq](https://api.jquery.com/eq/)

**Review by oliverfoster — APPROVED** [review]

**Review by cahirodoherty-learningpool — APPROVED** [review]
👀

**Comment by eleanor-heath** [comment]
@oliverfoster can you merge this please (I don't have the permissions to)

**Comment by oliverfoster** [comment]
@tomgreenfield needs to finish his review off with an approval

File: js/sliderView.js [hunk 12]
```diff
@@ -3,94 +3,91 @@ define([
...
-        this.$slider.prop('disabled', true).rangeslider('update', true);
-        return;
-      }
-
-      this.$('.slider__widget').removeClass('is-disabled');
-      this.$slider.prop('disabled', false).rangeslider('update', true);
-    },
+      this.$('.slider__widget').toggleClass('is-disabled', !isEnabled);
+      this.$slider.prop('disabled', !isEnabled).rangeslider('update', true);
+      return;
```
Comment (tomgreenfield) [inline js/sliderView.js:77]: Please remove the `return`.

File: js/sliderModel.js (hunk 7 above)
Comment (tomgreenfield) [reply js/sliderModel.js:106]: Using a filter would make things a lot simpler:
```js
isCorrect() {
  const numberOfCorrectAnswers = this.get('_items').filter(({ selected, correct }) => selected && correct).length;

  this.set('_isAtLeastOneCorrectSelection', numberOfCorrectAnswers > 0);
  this.set('_numberOfCorrectAnswers', numberOfCorrectAnswers);

  return this.get('_isAtLeastOneCorrectSelection');
}
```

File: js/sliderView.js [hunk 13]
```diff
@@ -289,102 +284,101 @@ define([
...
-
       this.setSliderValue(middleAnswer);
-    },
+    }
 
-    showModelAnswers: function(correctAnswerArray) {
-      var $parentDiv = this.$('.js-slider-model-range');
-      _.each(correctAnswerArray, function(correctAnswer, index) {
+    getCorrectAnswers() {
+      //are we dealing with a single correct answer or a range?
```
Comment (tomgreenfield) [inline js/sliderView.js:305]: Please add space before comment (linter will complain).

**Comment by eleanor-heath** [comment]
Hi @tomgreenfield, I've addressed all of your comments now. Now that I've managed to update to ES6 classes as well as do some general tidy-up, I think further refactors should be open in another tech-debt ticket.

Can you please re-review when you get a chance?

File: js/sliderView.js [hunk 14]
```diff
@@ -394,7 +388,7 @@ define([
       }
     }
 
-  });
+  };
```
Comment (oliverfoster) [inline js/sliderView.js:391]: Can you remove the `;` from the end of both classes? They're not necessary after a code block.
```js
if () {
}

class Name {
}

for () {
}
```

File: js/sliderView.js [hunk 15]
```diff
@@ -180,107 +176,105 @@ define([
...
+      const scaleWidth = $scaler.width();
+      this.model.get('_items').forEach((item, index) => {
+        const $number = $numbers.eq(index);
+        const newLeft = Math.round($number.data('normalisedPosition') * scaleWidth);
+        $number.css({ left: newLeft });
       });
-    },
+    }
 
     //Labels are enabled in slider.hbs. Here we manage their containing div.
```
Comment (oliverfoster) [inline js/sliderView.js:262]:
```suggestion
    // Labels are enabled in slider.hbs. Here we manage their containing div.
```

File: js/sliderView.js (hunk 14 above)
Comment (oliverfoster) [inline js/sliderView.js]:
```suggestion
  }
```

File: js/sliderModel.js [hunk 16]
```diff
@@ -70,84 +69,73 @@ define([
...
     /**
     * Used by adapt-contrib-spoor to get the type of this question in the format required by the cmi.interactions.n.type data field
     */
-    getResponseType:function() {
+    getResponseType() {
       return 'numeric';
     }
 
-  });
+  };
```
Comment (oliverfoster) [inline js/sliderModel.js]:
```suggestion
  }
```

File: js/sliderModel.js [hunk 17]
```diff
@@ -2,38 +2,37 @@ define([
...
 
-      for (var i = start; i <= end; i += step) {
+    setupModelItems() {
+      const items = [];
+      const answer = this.get('_correctAnswer');
+      const range = this.get('_correctRange');
+      const start = this.get('_scaleStart');
+      const end = this.get('_scaleEnd');
+      const step = this.get('_scaleStep') || 1;
+
```
Comment (oliverfoster) [inline js/sliderModel.js:32]:
```suggestion

```

File: js/sliderModel.js [hunk 18]
```diff
@@ -70,84 +69,73 @@ define([
...
       }
 
       this.setQuestionAsSubmitted();
       this.markQuestion();
       this.setScore();
       this.setupFeedback();
-    },
+    }
 
     //This preserves the state of the users answers for returning or showing the users answer
```
Comment (oliverfoster) [inline js/sliderModel.js:86]:
```suggestion
    // This preserves the state of the users answers for returning or showing the users answer
```

File: js/sliderModel.js [hunk 19]
```diff
@@ -70,84 +69,73 @@ define([
...
-          this.set('_isAtLeastOneCorrectSelection', true);
-          numberOfCorrectAnswers++;
-        }
-      }, this);
+    }
 
+    deselectAllItems() {
+      this.get('_items').forEach(item => item.selected = false);
+    }
+
```
Comment (oliverfoster) [inline js/sliderModel.js:102]:
```suggestion

```

File: js/sliderModel.js [hunk 20]
```diff
@@ -70,84 +69,73 @@ define([
...
-      }, this);
+    }
 
+    deselectAllItems() {
+      this.get('_items').forEach(item => item.selected = false);
+    }
+    
+    isCorrect() {
+      const numberOfCorrectAnswers = this.get('_items').filter(({ selected, correct }) => selected && correct).length;
+
```
Comment (oliverfoster) [inline js/sliderModel.js:105]:
```suggestion

```

File: js/sliderModel.js [hunk 21]
```diff
@@ -70,84 +69,73 @@ define([
...
+    deselectAllItems() {
+      this.get('_items').forEach(item => item.selected = false);
+    }
+    
+    isCorrect() {
+      const numberOfCorrectAnswers = this.get('_items').filter(({ selected, correct }) => selected && correct).length;
+    
+      this.set('_isAtLeastOneCorrectSelection', numberOfCorrectAnswers > 0);
       this.set('_numberOfCorrectAnswers', numberOfCorrectAnswers);
+
```
Comment (oliverfoster) [inline js/sliderModel.js:108]:
```suggestion

```

File: js/sliderModel.js [hunk 22]
```diff
@@ -2,38 +2,37 @@ define([
   'core/js/adapt',
   'core/js/models/questionModel'
 ], function(Adapt, QuestionModel) {
+  class SliderModel extends QuestionModel {
```
Comment (oliverfoster) [inline js/sliderModel.js:5]:
```suggestion

  class SliderModel extends QuestionModel {
```

**Review by oliverfoster — CHANGES_REQUESTED** [review]
Linter (ESLint plugin in VSCode) has some warnings.

If you're using VSCode it might be worth installing the ESLint plugin and adding these settings:
```json
{
  "editor.trimAutoWhitespace": true,
  "editor.renderWhitespace": "all",
  "files.trimTrailingWhitespace": true,
  "files.insertFinalNewline": true,
  "[markdown]": {
    "files.trimTrailingWhitespace": false
  }
}
```

File: js/sliderModel.js [hunk 23]
```diff
@@ -70,84 +69,73 @@ define([
...
-      _.each(this.get('_items'), function(item, index) {
-        if (item.selected && item.correct) {
-          this.set('_isAtLeastOneCorrectSelection', true);
-          numberOfCorrectAnswers++;
-        }
-      }, this);
+    }
 
+    deselectAllItems() {
+      this.get('_items').forEach(item => item.selected = false);
```
Comment (oliverfoster) [inline js/sliderModel.js:100]: Direct assignment returns aren't allowed
```suggestion
      this.get('_items').forEach(item => (item.selected = false));
```

**Comment by oliverfoster** [comment]
I reckon `getCorrectAnswers` should be a model function rather than a view function?

Change call to `this.model.getCorrectAnswers();`
```js
    showCorrectAnswer() {
      this.showScaleMarker(false);

      const answers = this.model.getCorrectAnswers();
      if (answers.length === 0) {
        console.log('adapt-contrib-slider::WARNING: no correct answer or correct range set in JSON');
        return;
      }

      const middleAnswer = answers[Math.floor(answers.length / 2)];
      this.animateToPosition(this.mapIndexToPixels(this.getIndexFromValue(middleAnswer)));
      this.showModelAnswers(answers);
      this.setSliderValue(middleAnswer);
    }
```

Move to model and change `this.model.` to `this.`
```js

    getCorrectAnswers() {

      // are we dealing with a single correct answer or a range?
      const answerSingle = this.get('_correctAnswer');
      const answers = [];
      if (answerSingle) {
        return [ answerSingle ];
      }
      const answerMultiple = this.get('_correctRange');
      if (!answerMultiple) {
        return answers;
      }
      const bottom = answerMultiple._bottom;
      const top = answerMultiple._top;
      if (bottom === undefined || top === undefined) {
        return answers;
      }
      let answer = bottom;
      const step = this.get('_scaleStep') || 1;
      while (answer <= top) {
        answers.push(answer);
        answer += step;
      }
      return answers;
    }
```

**Comment by eleanor-heath** [comment]
Hi @oliverfoster, i agree that these are good suggestions but would argue that they too are out of scope of this PR.

The depth of my changes were to introduce ES6 standards. Everything else that I have addressed was a quick tidy-up, but most of your suggestions are on code that is not necessary to change at this point.
If you would prefer, I can undo my tidy-up and only include the ES6 changes so that you can open a new PR to address any concerns that you may have.

**Comment by oliverfoster** [comment]
@eleanor-heath The linting problems break the project rules so you'll need to fix those. I have suggested above that you can modify your editor to help with those in future. Moving your new `getCorrectAnswers` function to the model won't take a minute to do and it is a model function.

The other stuff about items, indexes and `ItemComponentModel` is definitely out of scope, I agree.

**Review by oliverfoster — APPROVED** [review]

File: bower.json [hunk 24]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "adapt-contrib-slider",
-  "version": "3.0.0",
-  "framework": ">=5",
+  "version": "3.0.1",
```
Comment (moloko) [inline bower.json:3]: As this is more than just a 'bug fix'...
```suggestion
  "version": "3.1.0",
```

**Review by tomgreenfield — APPROVED** [review]
:eye:

**Pull request merged**